### ISSUE TITLE
Modified supervised contrastive loss with class-conditional InfoNCE

### ISSUE THREAD

**Issue by penguinwang96825** [KevinMusgrave/pytorch-metric-learning#573]
Hi, I saw an interesting paper called "Perfectly balanced: Improving transfer and robustness of supervised contrastive learning", which uses a so-called spread loss to robustly optimised the model. Simply speaking , the authors gave a weights on the supervised contrastive loss and class-conditional InfoNCE. I was just wondering if there's any way around to construct the class-conditioned InfoNCE without modifying the internal code of NTXent loss. Instead, with a few changes, we are able to get the class-conditional InfoNCE. Provide the GitHub [link](https://github.com/HazyResearch/thanos-code/) to the original [paper](https://arxiv.org/pdf/2204.07596.pdf)

**Comment by KevinMusgrave** [comment]
If I understand correctly, you want InfoNCE where the positive pair for a sample `x` is its augmentation `a(x)`, and the negative pairs are all samples with the same label as `x`, but excluding `a(x)`.

I think this will accomplish that:

```python
import torch
import numpy as np
from pytorch_metric_learning.losses import NTXentLoss
from pytorch_metric_learning.utils import loss_and_miner_utils as lmu

loss_fn = NTXentLoss()
embeddings = torch.randn(12, 128)

# I'll assume the second half of the batch is the augmented version of the first half
labels = torch.tensor([0, 0, 1, 1, 2, 2, 0, 0, 1, 1, 2, 2])
half_batch = len(labels)//2

# Create indices of each sample with its augmented version
# Convert it to a set of tuples
pos_pairs = set([(x, x+half_batch) for x in range(half_batch)])

# Get all positive pairs based on class labels
(all_a, all_p, _, _) = lmu.get_all_pairs_indices(labels)
all_pos_pairs = set([(x[0].item(),x[1].item()) for x in torch.stack([all_a, all_p], dim=1)])

# Negative pairs are the ones with the same class label
# but that are not in the pos_pairs that we constructed
neg_pairs = all_pos_pairs - pos_pairs

# convert to tensors
a1, p = torch.stack([torch.tensor(x) for x in pos_pairs], dim=1)
a2, n = torch.stack([torch.tensor(x) for x in neg_pairs], dim=1)
indices_tuple = (a1, p, a2, n)

# pass to loss_fn
loss = loss_fn(embeddings, indices_tuple=indices_tuple)
```
